contributors/design-proposals/hugepages.md

```diff
@@ -0,0 +1,281 @@
+eviction).  An application that consumes `hugepages.<hugepagesize>` is not in
+the `BestEffort` QoS class.
+
+If a pod consumes huge pages via `shmget`, it must run with a supplemental group
+that matches `/proc/sys/vm/hugetlb_shm_group` on the node.  Configuration of
+this group is outside the scope of this specification.
+
+In order to consume huge pages backed by the `hugetlbfs` filesystem inside the
+specified container in the pod, it is helpful to understand the set of mount
+options used with `hugetlbfs`.  For more details, see "Using Huge Pages" here:
+https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt
+
+```
+mount -t hugetlbfs \
+	-o uid=<value>,gid=<value>,mode=<value>,pagesize=<value>,size=<value>,\
+	min_size=<value>,nr_inodes=<value> none /mnt/huge
+```
+
+The proposal recommends a new volume type `hugePages`.  It will support a single
+`pageSize` option that must be provided by users.  The `pageSize` option aligns
+with the `pagesize` argument in the above mount.  It is used to specify the huge
+page size and pool that is used for accounting.  To write into this volume, the
+pod must make a request for huge pages that match this `pagesize`, otherwise its
+usage will be limited to 0 bytes.  No other option from the mount is carried
+over to the `hugePages` volume type.  The `size` option is defaulted to sum of
+`hugepages.<hugepagesize>` requests made by each container for the specified
+`pagesize`.  This keeps the behavior consistent with memory backed `emptyDir`
+volumes whose usage is ultimately constrained by the pod cgroup sandbox memory
+settings.  The `min_size` option is omitted as its not necessary. It's possible
+we could have extended `emptyDir` to support an additional option for
+`medium=Memory` but extend it to support a `pageSize` paramter, but this
+proposal prefers to create an additional volume type instead.
+
+The following is a sample pod that is limited to 512 huge pages of size 2MB. It
+can consume those pages using `shmget()` or via `mmap()` with the specified
+volume.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+...
+    volumeMounts:
+    - mountPath: /hugepages
+      name: hugepage
+    resources:
+      requests:
+        hugepages.2MB: "512"
+      limits:
+        hugepages.2MB: "512"
+  volumes:
+  - name: hugepage
+    hugePages:
+      pageSize: "2MB"
+```
+
+## CRI Updates
+
+The `LinuxContainerResources` message should be extended to support specifying
+huge page limits per size.  The specification for huge pages should align with
+opencontainers/runtime-spec.
+
+see:
+https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#huge-page-limits
+
+The CRI changes are required before promoting this feature to Beta.
+
+## Cgroup Enforcement
+
+To use this feature, the `--cgroups-per-qos` must be enabled.  In addition, the
+`hugetlb` cgroup must be mounted.
+
+The QoS level cgroups are left unbounded across all huge page pool sizes.
```

Comment:
do we want to bound at the kubepods (node) level in case system services are using hugepages?